#2012 Add support for helperText in all Common Properties controls.

srikant-ch5 opened the pull request:
Fixes: #2012
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

matthoward366 commented on canvas_modules/common-canvas/src/common-properties/controls/datepicker-range/datepicker-range.jsx:
```diff
@@ -152,7 +148,7 @@ class DatepickerRangeControl extends React.Component {
 						onChange={this.handleInputStartChange.bind(this)}
 						value={this.state.valueStart}
 						onBlur={this.onStartBlur.bind(this)}
-						helperText={!this.props.tableControl && startHelperText}
```
I don't think this is correct since there might be a different helper text for both.

srikant-ch5 commented:
Hi @matthoward366 should the pipeline schema be updated with helper-text ?.
In paramDef should helperText be defined as a string or an object with `default` just like placeHolderText.
In `datepicker.jsx` I have removed helperText being loaded from resource is this fine ?

matthoward366 commented on canvas_modules/common-canvas/src/common-properties/controls/datepicker/datepicker.jsx:
```diff
@@ -64,7 +64,6 @@ class DatepickerControl extends React.Component {
 	}
 
 	render() {
-		const helperText = this.props.controller.getResource(`${this.props.control.name}.helper`, null);
```
This would be a breaking change.  We might need to keep this but instead of having `null` using the control.helperText.

matthoward366 reviewed (commented):
Each control that supports helperText will need paramDef updates.

Can we also add some unit tests?

matthoward366 commented:
>In paramDef should helperText be defined as a string or an object with default just like placeHolderText.

Should be a resource object. Using a string isn't going to support multiple languages correctly.

>In datepicker.jsx I have removed helperText being loaded from resource is this fine ?

I added a comment about this.

srikant-ch5 commented:
Hi @matthoward366 What should be included in helper text as default value. Should it be control description or something like `control helper text`, should `helper_text` be also included in v3 schema ?

matthoward366 commented:
You should add it to the v3 schema.  There should be no default value.

srikant-ch5 commented:
> You should add it to the v3 schema. There should be no default value. it should pretty much match what placeholder_text has.

Sure, I have added the schema and have also updated this PR as per above feedback.

matthoward366 commented on canvas_modules/harness/test_resources/parameterDefs/checkbox_paramDef.json:
```diff
@@ -135,6 +135,9 @@
         "label": {
           "default": "checkbox default"
         },
+				"helper_text": {
```
Looks like the formatting is off

srikant-ch5 replied:
I have formatted this file using prettier is this fine or should I revert it and only format my changes ?

matthoward366 replied:
You can format the entire file.  Just make sure it using the standard json formatting.

srikant-ch5 replied:
Sure. I have updated accordingly.

matthoward366 commented on canvas_modules/harness/test_resources/parameterDefs/datepicker_paramDef.json:
```diff
@@ -101,8 +101,11 @@
 				"place_holder_text": {
 					"default": "Enter a date: mm-dd-yyyy"
+				},
+				"helper_text": {
```
formatting?

matthoward366 reviewed (commented):
I don't think we need this file `canvas_modules/common-canvas/podman`

Just a few minor items.

matthoward366 commented:
helperText is missing in checkboxset which we should wrap in a `CheckboxGroup`.  This can be done in a new issue.  Can you create it?

matthoward366 reviewed (approved):
Looks good

srikant-ch5 commented:
> helperText is missing in checkboxset which we should wrap in a `CheckboxGroup`. This can be done in a new issue. Can you create it?

Created an issue regarding the same #3811.
